### PR TITLE
WIP: Add eTag support to API discovery

### PR DESCRIPTION
**Pull request by mbohlool**
Still work in progress. Created PR to show the direction and discuss it.

ref: #47977 https://github.com/kubernetes/kubernetes/issues/44957

### CODE REVIEW

**Comment by apelisse** [comment]
LGTM

**Comment by mbohlool** [comment]
I am not sure if I want to call this ready yet. Need a little more time to write a simple design for this. But I would be soon. Do you guys have any deadline that you need this? (if you need it soon I can make this PR ready to merge for now and add other stuffs later).

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go [hunk 1]
```diff
@@ -165,3 +168,16 @@ func WriteRawJSON(statusCode int, object interface{}, w http.ResponseWriter) {
...
+
+// WriteRawJSON writes a non-API object in JSON with eTag support.
+func WriteRawJSONWithETag(object interface{}, w http.ResponseWriter, r *http.Request) {
+	output, err := json.MarshalIndent(object, "", "  ")
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Etag", fmt.Sprintf("\"%X\"", sha512.Sum512(output)))
+	w.Header().Set("Content-Type", "application/json")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go]: I'd expect a cache control header set here as well

Comment (mbohlool) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go]: Do we assume all cache controls should be "private"?

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go]: I would start there, yeah

**Pull request closed without merging**